### PR TITLE
sqlbase: add limited prefetch to txnKVFetcher (WIP)

### PR DESCRIPTION
**Pull request by jordanlewis**
This is a WIP - it's currently got a bug. Just pushing up for people to take a look at early.

Previously, the txnKVFetcher was single-threaded and waited for its
invoker to call fetch on it to do any work.

Now, after the first call to fetch, the next fetch is always initiated
in the background.

This seems to improve performance by about 10% on COUNT(*) locally. The
reasoning is that the IO that the fetcher initiates shouldn't be going
on in the foreground if there is actual CPU work to be done at that
point.

Release note: None

### CODE REVIEW

**Comment by petermattis** [comment]
Review status: 0 of 1 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[pkg/sql/sqlbase/kvfetcher.go, line 391 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21773#-L3ew8Ejj_v0mVXrD7Dt:-L3ew8Ejj_v0mVXrD7Du:bwtljx2) ([raw file](https://github.com/cockroachdb/cockroach/blob/d1dcf24ae5abece037a1b9c4c42a62303bea5aa7/pkg/sql/sqlbase/kvfetcher.go#L391)):*
> ```Go
> 	}
> 
> 	if !f.fetchEnd {
> ```

This heuristic is fine for experimentation, but probably a pessimization in practice. IIUC, you're reading the next batch of rows as soon as you finish reading the previous batch. I think a more realistic heuristic is to start fetching the next batch of rows when the previous batch is almost exhausted. That is, kick of this async fetch in `nextKV()` by looking at `len(f.kvs)`.

Cc @raduberinde as he might have thoughts about this as well.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21773)*

File: pkg/sql/sqlbase/kvfetcher.go [hunk 1]
```diff
@@ -22,6 +22,8 @@ import (
 
 	"github.com/pkg/errors"
 
+	"sync"
```
Comment (benesch) [inline pkg/sql/sqlbase/kvfetcher.go]: this import ended up in a weird place

**Comment by RaduBerinde** [comment]
Make sure to run the benchmarks that have LIMIT. We may be making some of those a lot worse.

---

Review status: 0 of 1 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[pkg/sql/sqlbase/kvfetcher.go, line 286 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21773#-L3hz5dxwKXv9BsCZ8nY:-L3hz5dxwKXv9BsCZ8nZ:b-gaxvp6) ([raw file](https://github.com/cockroachdb/cockroach/blob/d1dcf24ae5abece037a1b9c4c42a62303bea5aa7/pkg/sql/sqlbase/kvfetcher.go#L286)):*
> ```Go
> }
> 
> func (f *txnKVFetcher) processBatch(ctx context.Context) error {
> ```

a comment would be good here

---

*[pkg/sql/sqlbase/kvfetcher.go, line 391 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21773#-L3ew8Ejj_v0mVXrD7Dt:-L3hxtaozG6iqSloMV13:b-xeu43l) ([raw file](https://github.com/cockroachdb/cockroach/blob/d1dcf24ae5abece037a1b9c4c42a62303bea5aa7/pkg/sql/sqlbase/kvfetcher.go#L391)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

This heuristic is fine for experimentation, but probably a pessimization in practice. IIUC, you're reading the next batch of rows as soon as you finish reading the previous batch. I think a more realistic heuristic is to start fetching the next batch of rows when the previous batch is almost exhausted. That is, kick of this async fetch in `nextKV()` by looking at `len(f.kvs)`.

Cc @raduberinde as he might have thoughts about this as well.
</blockquote></details>

I like the suggestion, we could experiment with the threshold (as a percentage). The threshold could even be different depending if we have a limit (or a soft limit vs a hard limit).

I don't quite get what the code is doing here (or I might be misunderstanding the change). It seems that we get batches in pairs. We will kick off the first batch, then the second batch right away. Then we have to consume everything before we start the third batch (which is followed by batch 4 immediately). And so on.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21773#-:-L3hxq4IzuiDC6S6neXB:b-kypbhm)*

**Comment by jordanlewis** [comment]
Review status: 0 of 1 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[pkg/sql/sqlbase/kvfetcher.go, line 391 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21773#-L3ew8Ejj_v0mVXrD7Dt:-L3i4n0JIuJrjQLl2NKr:b-7qa53z) ([raw file](https://github.com/cockroachdb/cockroach/blob/d1dcf24ae5abece037a1b9c4c42a62303bea5aa7/pkg/sql/sqlbase/kvfetcher.go#L391)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

I like the suggestion, we could experiment with the threshold (as a percentage). The threshold could even be different depending if we have a limit (or a soft limit vs a hard limit).

I don't quite get what the code is doing here (or I might be misunderstanding the change). It seems that we get batches in pairs. We will kick off the first batch, then the second batch right away. Then we have to consume everything before we start the third batch (which is followed by batch 4 immediately). And so on.
</blockquote></details>

I think you're misunderstanding (or maybe I made a mistake). Every call to `fetch` hits `processBatch`, which has two paths. If there's no prefetched data, it does a synchronous fetch. If there's prefetched data, it immediately returns that prefetched data.

In both cases, if the fetch isn't finished, we kick off an asynchronous prefetch right before returning from `fetch`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21773)*

**Comment by RaduBerinde** [comment]
*[pkg/sql/sqlbase/kvfetcher.go, line 391 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21773#-L3ew8Ejj_v0mVXrD7Dt:-L3iugPlviy6wEKmwwbi:b-qfaans) ([raw file](https://github.com/cockroachdb/cockroach/blob/d1dcf24ae5abece037a1b9c4c42a62303bea5aa7/pkg/sql/sqlbase/kvfetcher.go#L391)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

I think you're misunderstanding (or maybe I made a mistake). Every call to `fetch` hits `processBatch`, which has two paths. If there's no prefetched data, it does a synchronous fetch. If there's prefetched data, it immediately returns that prefetched data.

In both cases, if the fetch isn't finished, we kick off an asynchronous prefetch right before returning from `fetch`.
</blockquote></details>

Ah, I see, thanks!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21773)*

**Comment by jordanlewis** [comment]
cc @solongordon

**Comment by jordanlewis** [comment]
Closing this though I think we'll want to pursue something similar one day.

**Pull request closed without merging**